salt/utils/vmware.py

```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
+                                             'parameter \'password\' is '
+                                             'missing')
+
+        if domain and '@' not in username and '\\' not in username:
+            # Append the domain if the username doesn't already contain a
+            # domain (either old or new style representation)
+            username = '{0}@{1}'.format(username, domain)
+            log.debug('Domain added to username: \'{0}\''.format(username))
+    elif mechanism == 'sspi':
+        if principal is not None and domain is not None:
+            try:
+                token = get_gssapi_token(principal, host, domain)
+            except Exception as exc:
+                raise excs.VMwareConnectionError(str(exc))
+        else:
+            err_msg = 'Login mechanism \'{0}\' was specified but the' \
+                      ' mandatory parameters are missing'.format(mechanism)
+            raise excs.CommandExecutionError(err_msg)
+    else:
+        raise excs.CommandExecutionError('Unsupported mechanism: \'{0}\''
+                                         ''.format(mechanism))
     try:
+        log.debug('Connecting using the \'{0}\' mechanism, with username '
+                  '\'{1}\''.format(mechanism, username))
         service_instance = SmartConnect(
             host=host,
             user=username,
             pwd=password,
             protocol=protocol,
-            port=port
-        )
+            port=port,
+            b64token=token,
+            mechanism=mechanism)
     except Exception as exc:
         default_msg = 'Could not connect to host \'{0}\'. ' \
                       'Please check the debug log for more information.'.format(host)
         try:
-            if (isinstance(exc, vim.fault.HostConnectFault) and '[SSL: CERTIFICATE_VERIFY_FAILED]' in exc.msg) or '[SSL: CERTIFICATE_VERIFY_FAILED]' in str(exc):
-                import ssl
-                default_context = ssl._create_default_https_context
-                ssl._create_default_https_context = ssl._create_unverified_context
-                service_instance = SmartConnect(
-                    host=host,
-                    user=username,
-                    pwd=password,
-                    protocol=protocol,
-                    port=port
-                )
-                ssl._create_default_https_context = default_context
-            elif (isinstance(exc, vim.fault.HostConnectFault) and 'SSL3_GET_SERVER_CERTIFICATE\', \'certificate verify failed' in exc.msg) or 'SSL3_GET_SERVER_CERTIFICATE\', \'certificate verify failed' in str(exc):
+            if (isinstance(exc, vim.fault.HostConnectFault) and
+                '[SSL: CERTIFICATE_VERIFY_FAILED]' in exc.msg) or \
+               '[SSL: CERTIFICATE_VERIFY_FAILED]' in str(exc):
+
                 import ssl
-                default_context = ssl._create_default_https_context
-                ssl._create_default_https_context = ssl._create_unverified_context
```

Comment:
I tested connections both with unverified and verified contexts, but with pyVmomi 6.0. I can't test with older versions unfortunately